### PR TITLE
refactor(core/utils): use Intl.ListFormat for joining

### PR DESCRIPTION
**Pull request by marcoscaceres**

### CODE REVIEW

File: src/type-helper.d.ts [hunk 1]
```diff
@@ -81,6 +81,10 @@ declare function fetch(input: URL, init?: RequestInit): Promise<Response>;
 
 declare namespace Intl {
   class ListFormat {
+    formatToParts(items: string): {
+      type: "element" | "literal";
+      value: string;
+    };
```
Comment (sidvishnoi) [inline src/type-helper.d.ts:87]:
```suggestion
    }[];
```

Comment (sidvishnoi) [inline src/type-helper.d.ts:87]: This should remove the need for ts-ignore in `makeListJoiner`
```suggestion
    }[];
```

**Review by sidvishnoi — COMMENTED** [review]
I like where this is going! Let's add TS-types to all the new utils though..

File: src/core/utils.js [hunk 2]
```diff
@@ -104,41 +92,54 @@ function markAsOffending(elem, msg, title) {
...
+   * @param {(value: string, index: number, array: string[]) => any} [mapper]
+   */
+  return (items, mapper) => {
+    let elemCount = 0;
+    if (!mapper) {
+      return [formatter.format(items)];
+    }
+    return formatter.formatToParts(items).map(({ type, value }) => {
+      if (type === "element") {
+        return mapper(value, elemCount++, items);
```
Comment (sidvishnoi) [inline src/core/utils.js:112]: Looks like we don't use it.
```suggestion
        return mapper(value, elemCount++);
```

File: src/core/utils.js [hunk 3]
```diff
@@ -104,41 +92,54 @@ function markAsOffending(elem, msg, title) {
...
+ * @param {"conjunction"|"disjunction"} type
+ * @param {"long"|"narrow"} style
+ */
+function joinFactory(type, style = "long") {
+  const formatter = new Intl.ListFormat(docLang, { style, type });
+  /**
+   * @param {string[]} items
+   * @param {(value: string, index: number, array: string[]) => any} [mapper]
+   */
+  return (items, mapper) => {
```
Comment (sidvishnoi) [inline src/core/utils.js:105]: How about something like:
```ts
return (items: T[], mapper?: (item: T) => any), getter?: (item: T) => string) => {}
```

File: src/core/utils.js (hunk 2 above)
Comment (marcoscaceres) [reply src/core/utils.js:113]: I'd rather sent the interface compliant map.

Comment (marcoscaceres) [reply src/core/utils.js:113]: (even though we don't use it)

File: src/core/utils.js (hunk 3 above)
Comment (marcoscaceres) [reply src/core/utils.js:107]: Added.

**Comment by marcoscaceres** [comment]
hopefully all good now.

File: src/core/utils.js (hunk 3 above)
Comment (sidvishnoi) [reply src/core/utils.js:107]: I meant:
```diff
-   return (items, mapper) => { 
+   return (items: T[], mapper?: (item: T) => any), getter?: (item: T) => string) => {
+   // above, but in JS (not TS)
```

File: src/core/utils.js [hunk 4]
```diff
@@ -104,41 +92,56 @@ function markAsOffending(elem, msg, title) {
...
+ * @param {"conjunction"|"disjunction"} type
+ * @param {"long"|"narrow"} style
+ */
+function joinFactory(type, style = "long") {
+  const formatter = new Intl.ListFormat(docLang, { style, type });
+  /**
+   * @template T
+   * @param {string[]} items
+   * @param {(value: string, index: number, array: string[]) => any} [mapper]
+   * @return (items: T[], mapper?: (item: T) => any), getter?: (item: T) => string) => {}
```
Comment (sidvishnoi) [inline src/core/utils.js:105]: Let's ignore this for now. We can make it more generic next time we need it.

```suggestion
```

File: src/core/utils.js (hunk 3 above)
Comment (sidvishnoi) [reply src/core/utils.js:106]: Let's ignore this for now. We can make it more generic next time we need it.

File: src/core/utils.js [hunk 5]
```diff
@@ -104,41 +92,55 @@ function markAsOffending(elem, msg, title) {
...
+    let elemCount = 0;
+    if (!mapper) {
+      return [formatter.format(items)];
+    }
+    return formatter.formatToParts(items).map(({ type, value }) => {
+      if (type === "element") {
+        return mapper(value, elemCount++, items);
+      }
+      return value;
+    });
```
Comment (sidvishnoi) [inline src/core/utils.js:116]:
```suggestion
    return formatter.formatToParts(items).map(({ type, value }) => {
      if (type === "element" && mapper) {
        return mapper(value, elemCount++, items);
      }
      return value;
    });
```

File: src/core/utils.js (hunk 3 above)
Comment (sidvishnoi) [reply src/core/utils.js:106]: Actually maybe we can?

**Review by sidvishnoi — APPROVED** [review]
Looks good.
Would like to remove the "must pass string[] to `joinAnd`" requirement in future, and make the interface more like `Array.map`.

**Pull request merged**